lib/inspec/input_registry.rb

```diff
@@ -166,15 +166,44 @@ def bind_inputs_from_cli_args(profile_name, input_list)
           end
         end
         input_name, input_value = pair.split("=")
+        input_value = clean_up_cli_input_value(input_name, input_value)
         evt = Inspec::Input::Event.new(
-          value: input_value.chomp(","), # Trim trailing comma if any
+          value: input_value,
           provider: :cli,
           priority: 50
         )
         find_or_register_input(input_name, profile_name, event: evt)
       end
     end
 
+    # Remove trailing commas, resolve type.
+    def clean_up_cli_input_value(input_name, given_value)
+      value = given_value.chomp(",") # Trim trailing comma if any
+      if value =~ /^-?\d+$/
+        value = value.to_i
+      elsif value =~ /^-?\d+\.\d+$/
+        value = value.to_f
+      elsif value =~ /^(\[|\{).*(\]|\})$/
+        # Look for complex values and try to parse them.
+        require "yaml"
+        begin
+          value = YAML.load(value)
+        rescue Psych::SyntaxError => yaml_error
+          # It could be that we just tried to run JSON through the YAML parser.
+          require "json"
+          begin
+            value = JSON.parse(value)
+          rescue JSON::ParserError => json_error
+            msg = "ERROR: Unparseable value '#{value}' for --input #{input_name}.\n"
+            msg += "When treated as YAML, error: #{yaml_error.message}\n"
+            msg += "When treated as JSON, error: #{json_error.message}"
```

Comment:
I like this approach/flow all the way down to showing possible errors for yaml/json.

Normally I'd be the first to ask conditionals to be named and refactored, along with nested begin/rescues, but in this case I actually think this design is the most straightforward/readable- nice!